File: 7_Visitors/collision.cpp
#include <iostream>

struct Object {};
struct Car: public Object {int x;};
struct Wall: public Object {int y;};

void _collide(Car* c, Car* w) {
  std::cout << "car" << " wall\n";
}

void _collide(Car* c, Wall* w) {
  std::cout << "car" << " wall\n";
}

void _collide(Wall* c, Car* w) {
  std::cout << "car" << " wall\n";
}

void _collide(Wall* c, Wall* w) {
  std::cout << "car" << " wall\n";
}

void collide(Object* o0, Object* o1) {
  std::cout << "Calling dispatch\n";
  _collide(o0, o1);
}

int main() {
  struct Object* o0 = new Car();
  struct Object* o1 = new Wall();
  collide(o0, o1);
}

File: 7_Visitors/collision.jl
abstract type Object end

struct Car <: Object end
struct Wall <: Object end

function collide(a::Object, b::Object)
    println("Calling dispatch")
    _collide(a, b)
end

_collide(a::Car, b::Car) = println("Two cars collided")
_collide(a::Car, b::Wall) = println("Car hit wall")
_collide(a::Wall, b::Car) = println("Wall hit car")
_collide(a::Wall, b::Wall) = println("Two walls collided")

collide(Car(), Wall())
